File: src/shared/types/dto/activityPanel.ts
import type {
  ActivityPanelDocument,
  ActivityPanelRow,
  ActivityPanelSnapshot,
  BackendActivityPanelId,
} from "../domain/activityPanel";
import { isNodeCreationDescriptorDto } from "../domain/nodeCreationDescriptor";

function record(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}
function exact(value: Record<string, unknown>, keys: readonly string[]): boolean {
  return (
    Object.keys(value).length === keys.length &&
    keys.every((key) => Object.prototype.hasOwnProperty.call(value, key))
  );
}
function text(value: unknown): boolean {
  return (
    record(value) &&
    ((exact(value, ["key"]) && typeof value.key === "string") ||
      (exact(value, ["text"]) && typeof value.text === "string"))
  );
}
function integer(value: unknown): value is number {
  return typeof value === "number" && Number.isSafeInteger(value) && value >= 0;
}
const actions = ["install", "refresh"];
function tools(value: unknown): boolean {
  return (
    Array.isArray(value) &&
    value.length <= 8 &&
    new Set(value.map((tool) => (record(tool) ? tool.id : null))).size === value.length &&
    value.every(
      (tool) =>
        record(tool) &&
        exact(tool, ["id", "label", "icon"]) &&
        actions.includes(tool.id as string) &&
        ["add", "install", "refresh"].includes(tool.icon as string) &&
        text(tool.label),
    )
  );
}
function item(value: unknown): boolean {
  if (!record(value)) return false;
  const strings = (keys: string[]) => keys.every((key) => typeof value[key] === "string");
  switch (value.kind) {
    case "node":
      return (
        exact(value, ["kind", "key", "title", "creation"]) &&
        strings(["key", "title"]) &&
        isNodeCreationDescriptorDto(value.creation)
      );
    case "command":
      return (
        exact(value, ["kind", "id", "label"]) &&
        ["undo", "redo"].includes(value.id as string) &&
        text(value.label)
      );
    case "plugin":
      return (
        exact(value, ["kind", "id", "name", "description", "publisher", "enabled"]) &&
        strings(["id", "name", "description", "publisher"]) &&
        typeof value.enabled === "boolean"
      );
    default:
      return false;
  }
}

export function parseActivityPanelDocument(value: unknown): ActivityPanelDocument | null {
  if (
    !record(value) ||
    !exact(value, [
      "schema",
      "panelId",
      "projectInstanceId",
      "publicationRevision",
      "title",
      "tools",
      "rows",
      "emptyState",
    ])
  )
    return null;
  if (
    value.emptyState !== null &&
    (!record(value.emptyState) ||
      !exact(value.emptyState, ["title", "description"]) ||
      !text(value.emptyState.title) ||
      !text(value.emptyState.description))
  )
    return null;
  if (
    value.schema !== "yssbi.activity-panel.v1" ||
    !["nodes", "commands", "plugins"].includes(value.panelId as string)
  )
    return null;
  if (value.projectInstanceId !== null && typeof value.projectInstanceId !== "string") return null;
  if (!integer(value.publicationRevision) || !text(value.title) || !tools(value.tools)) return null;
  if (!Array.isArray(value.rows) || value.rows.length > 10_000) return null;
  const ids = new Set<string>();
  let previousDepth = -1;
  let previousCategory = true;
  for (const row of value.rows) {
    if (
      !record(row) ||
      typeof row.id !== "string" ||
      !row.id ||
      ids.has(row.id) ||
      !integer(row.depth) ||
      row.depth > 32
    )
      return null;
    if (row.depth > previousDepth && (row.depth !== previousDepth + 1 || !previousCategory))
      return null;
    previousDepth = row.depth;
    previousCategory = row.kind === "category";
    ids.add(row.id);
    switch (row.kind) {
      case "category":
        if (
          !exact(row, ["id", "depth", "kind", "label", "defaultExpanded", "tools", "count"]) ||
          !text(row.label) ||
          typeof row.defaultExpanded !== "boolean" ||
          !tools(row.tools) ||
          !(row.count === null || integer(row.count))
        )
          return null;
        break;
      case "item":
        if (!exact(row, ["id", "depth", "kind", "item"]) || !item(row.item)) return null;
        if (
          !record(row.item) ||
          !{
            nodes: ["node"],
            commands: ["command"],
            plugins: ["plugin"],
          }[value.panelId as BackendActivityPanelId].includes(row.item.kind as string)
        )
          return null;
        break;
      case "message":
        if (
          !exact(row, ["id", "depth", "kind", "label", "description"]) ||
          !text(row.label) ||
          !(row.description === null || text(row.description))
        )
          return null;
        break;
      default:
        return null;
    }
  }
  return value as unknown as ActivityPanelDocument;
}

function cursor(value: unknown): value is string {
  return typeof value === "string" && value.length > 0 && value.length <= 64;
}

const HEADER_PATCH_FIELDS = new Set(["title", "tools", "emptyState", "publicationRevision"]);
const ROW_PATCH_FIELDS = new Set([
  "depth",
  "label",
  "defaultExpanded",
  "tools",
  "count",
  "item",
  "description",
]);

/** Apply wire operations to an isolated copy; a rejected batch never mutates the visible projection. */
export function parseActivityPanelUpdate(
  value: unknown,
  previous: ActivityPanelSnapshot | null,
): ActivityPanelSnapshot | null {
  if (!record(value) || !cursor(value.cursor)) return null;
  if (value.kind === "snapshot") {
    if (!exact(value, ["kind", "cursor", "document"])) return null;
    const document = parseActivityPanelDocument(value.document);
    return document ? { cursor: value.cursor, document } : null;
  }
  if (
    value.kind !== "patch" ||
    !exact(value, ["kind", "baseCursor", "cursor", "patch", "operations"]) ||
    !previous ||
    value.baseCursor !== previous.cursor
  )
    return null;
  if (!record(value.patch) || Object.keys(value.patch).some((key) => !HEADER_PATCH_FIELDS.has(key)))
    return null;
  if (!Array.isArray(value.operations) || value.operations.length > 30_000) return null;
  if (value.operations.length === 0 && Object.keys(value.patch).length === 0) {
    return value.cursor === previous.cursor
      ? previous
      : { cursor: value.cursor, document: previous.document };
  }
  if (value.cursor === previous.cursor) return null;

  const rows: unknown[] = value.operations.length ? [...previous.document.rows] : [];
  let indices: Map<string, number> | null = null;
  const find = (id: string) => {
    indices ??= new Map(rows.map((row, index) => [(row as ActivityPanelRow).id, index]));
    return indices.get(id) ?? -1;
  };
  for (const operation of value.operations) {
    if (!record(operation)) return null;
    if (operation.op === "insert") {
      if (
        !exact(operation, ["op", "afterId", "row"]) ||
        !record(operation.row) ||
        typeof operation.row.id !== "string" ||
        !operation.row.id ||
        find(operation.row.id) !== -1
      )
        return null;
      if (operation.afterId !== null && typeof operation.afterId !== "string") return null;
      const after = operation.afterId === null ? -1 : find(operation.afterId);
      if (operation.afterId !== null && after < 0) return null;
      rows.splice(after + 1, 0, operation.row);
      indices = null;
      continue;
    }
    if (typeof operation.id !== "string") return null;
    const index = find(operation.id);
    if (index < 0) return null;
    switch (operation.op) {
      case "update":
        if (
          !exact(operation, ["op", "id", "patch"]) ||
          !record(operation.patch) ||
          Object.keys(operation.patch).length === 0 ||
          Object.keys(operation.patch).some((key) => !ROW_PATCH_FIELDS.has(key))
        )
          return null;
        rows[index] = { ...(rows[index] as ActivityPanelRow), ...operation.patch };
        break;
      case "remove":
        if (!exact(operation, ["op", "id"])) return null;
        rows.splice(index, 1);
        indices = null;
        break;
      case "move": {
        if (
          !exact(operation, ["op", "id", "afterId"]) ||
          operation.afterId === operation.id ||
          (operation.afterId !== null && typeof operation.afterId !== "string")
        )
          return null;
        const [row] = rows.splice(index, 1);
        indices = null;
        const after = operation.afterId === null ? -1 : find(operation.afterId);
        if (operation.afterId !== null && after < 0) return null;
        rows.splice(after + 1, 0, row);
        indices = null;
        break;
      }
      default:
        return null;
    }
  }
  const document = parseActivityPanelDocument({
    ...previous.document,
    ...value.patch,
    rows: value.operations.length ? rows : previous.document.rows,
  });
  if (!document || document.publicationRevision < previous.document.publicationRevision)
    return null;
  return { cursor: value.cursor, document };
}

File: src-tauri/crates/yss-api/src/schema/activity_panel.rs
use serde::Serialize;
use yss_application::activity_panel as app;

use super::catalog::NodeCreationDescriptorDto;
use crate::error::CommandError;

#[derive(Debug, Serialize)]
#[serde(untagged)]
pub enum ActivityTextDto {
    Key { key: &'static str },
    Literal { text: String },
}
impl From<app::ActivityText> for ActivityTextDto {
    fn from(text: app::ActivityText) -> Self {
        match text {
            app::ActivityText::Key(key) => Self::Key { key },
            app::ActivityText::Literal(text) => Self::Literal { text },
        }
    }
}

#[derive(Debug, Serialize)]
pub struct ActivityToolDto {
    id: &'static str,
    label: ActivityTextDto,
    icon: &'static str,
}
impl From<app::ActivityTool> for ActivityToolDto {
    fn from(tool: app::ActivityTool) -> Self {
        Self {
            id: tool.id,
            label: tool.label.into(),
            icon: tool.icon,
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(
    tag = "kind",
    rename_all = "camelCase",
    rename_all_fields = "camelCase"
)]
pub enum ActivityItemDto {
    Node {
        key: String,
        title: String,
        creation: NodeCreationDescriptorDto,
    },
    Command {
        id: &'static str,
        label: ActivityTextDto,
    },
    Plugin {
        id: String,
        name: String,
        description: String,
        publisher: String,
        enabled: bool,
    },
}
impl From<app::ActivityItem> for ActivityItemDto {
    fn from(item: app::ActivityItem) -> Self {
        match item {
            app::ActivityItem::Node {
                key,
                title,
                creation,
            } => Self::Node {
                key,
                title,
                creation: creation.into(),
            },
            app::ActivityItem::Command { id, label } => Self::Command {
                id,
                label: label.into(),
            },
            app::ActivityItem::Plugin {
                id,
                name,
                description,
                publisher,
                enabled,
            } => Self::Plugin {
                id,
                name,
                description,
                publisher,
                enabled,
            },
        }
    }
}

#[derive(Debug, Serialize)]
#[serde(
    tag = "kind",
    rename_all = "camelCase",
    rename_all_fields = "camelCase"
)]
pub enum ActivityRowContentDto {
    Category {
        label: ActivityTextDto,
        default_expanded: bool,
        tools: Vec<ActivityToolDto>,
        count: Option<usize>,
    },
    Item {
        item: ActivityItemDto,
    },
    Message {
        label: ActivityTextDto,
        description: Option<ActivityTextDto>,
    },
}
#[derive(Debug, Serialize)]
pub struct ActivityRowDto {
    id: String,
    depth: usize,
    #[serde(flatten)]
    content: ActivityRowContentDto,
}

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ActivityPanelDocumentDto {
    schema: &'static str,
    panel_id: &'static str,
    project_instance_id: Option<String>,
    publication_revision: u64,
    title: ActivityTextDto,
    tools: Vec<ActivityToolDto>,
    rows: Vec<ActivityRowDto>,
    empty_state: Option<ActivityEmptyStateDto>,
}

#[derive(Debug, Serialize)]
struct ActivityEmptyStateDto {
    title: ActivityTextDto,
    description: ActivityTextDto,
}

impl TryFrom<app::ActivityPanelDocument> for ActivityPanelDocumentDto {
    type Error = CommandError;
    fn try_from(document: app::ActivityPanelDocument) -> Result<Self, Self::Error> {
        if document.rows.len() > 10_000 || document.rows.iter().any(|row| row.depth > 32) {
            return Err(CommandError::expected("activity_panel_limit_exceeded"));
        }
        let document = Self {
            schema: "yssbi.activity-panel.v1",
            panel_id: document.panel_id,
            project_instance_id: document.project_instance_id,
            publication_revision: document.publication_revision,
            title: document.title.into(),
            tools: document.tools.into_iter().map(Into::into).collect(),
            empty_state: document
                .empty_state
                .map(|(title, description)| ActivityEmptyStateDto {
                    title: title.into(),
                    description: description.into(),
                }),
            rows: document
                .rows
                .into_iter()
                .map(|row| ActivityRowDto {
                    id: row.id,
                    depth: row.depth,
                    content: match row.content {
                        app::ActivityRowContent::Category {
                            label,
                            default_expanded,
                            tools,
                            count,
                        } => ActivityRowContentDto::Category {
                            label: label.into(),
                            default_expanded,
                            tools: tools.into_iter().map(Into::into).collect(),
                            count,
                        },
                        app::ActivityRowContent::Item(item) => {
                            ActivityRowContentDto::Item { item: item.into() }
                        }
                        app::ActivityRowContent::Message { label, description } => {
                            ActivityRowContentDto::Message {
                                label: label.into(),
                                description: description.map(Into::into),
                            }
                        }
                    },
                })
                .collect(),
        };
        Ok(document)
    }
}
